pkg/cluster/fixetcd.go

```diff
@@ -0,0 +1,499 @@
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	securityv1 "github.com/openshift/api/security/v1"
+	operatorv1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+	securityv1client "github.com/openshift/client-go/security/clientset/versioned/typed/security/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/ugorji/go/codec"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	batchv1client "k8s.io/client-go/kubernetes/typed/batch/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+)
+
+const (
+	serviceAccountName                  = "etcd-recovery-privileged"
+	kubeServiceAccount                  = "system:serviceaccount" + nameSpaceEtcds + ":" + serviceAccountName
+	nameSpaceEtcds                      = "openshift-etcd"
+	image                               = "ubi8/ubi-minimal"
+	jobName                             = "etcd-recovery-"
+	jobNameDataBackup                   = jobName + "data-backup"
+	jobNameFixPeers                     = jobName + "fix-peers"
+	patchOverides                       = "unsupportedConfigOverrides:"
+	patchDisableOverrides               = `{"useUnsupportedUnsafeNonHANonProductionUnstableEtcd": true}`
+	ctxKey                timeoutAdjust = "TESTING"
+)
+
+type timeoutAdjust string
+
+type degradedEtcd struct {
+	Node  string
+	Pod   string
+	NewIP string
+	OldIP string
+}
+
+func (m *manager) fixEtcd(ctx context.Context) error {
+	pods, err := m.kubernetescli.CoreV1().Pods(nameSpaceEtcds).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	// Exit early if degradedEtcd is empty, an error is returned
+	// no further actions should be taken
+	de, err := comparePodEnvToIP(m.log, pods)
+	if err != nil {
+		return err
+	}
+	m.log.Infof("Found degraded endpoint: %v", de)
+
+	err = backupEtcdData(ctx, m.log, de.Node, m.kubernetescli.BatchV1())
+	if err != nil {
+		return err
+	}
+
+	err = fixPeers(ctx, m.log, de, pods, m.kubernetescli, m.securitycli.SecurityV1().SecurityContextConstraints(), m.doc.OpenShiftCluster.Name)
+	if err != nil {
+		return err
+	}
+
+	etcd, err := m.operatorcli.OperatorV1().Etcds().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	etcd.Spec.UnsupportedConfigOverrides = kruntime.RawExtension{
+		Raw: []byte(patchDisableOverrides),
+	}
+	err = patchEtcd(ctx, m.log, m.operatorcli.OperatorV1().Etcds(), etcd, patchDisableOverrides)
+	if err != nil {
+		return err
+	}
```

Comment:
@SudoBrendan do you mean the unsupportedConfigOverrides that we're patching in here? If so, we are just using the official docs https://docs.openshift.com/container-platform/4.11/backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.html as a guide here.

Reply:
@SudoBrendan I updated this part to save the existing overrides, and then reassign the overrides (which maybe nil) rather than just resetting it to nil.